### PR TITLE
CloudWatch: replace metricFindQueries with CallResourceHandler

### LINKED ISSUE #40748: CloudWatch: Replace MetricFindQueries with CallResourceHandler
### Discussed in https://github.com/grafana/grafana/discussions/40747

<div type='discussions-op-text'>

<sup>Originally posted by **sunker** October 21, 2021</sup>
Currently, variable queries and queries for populating the query editor with meta data are using the `QueryData` api in the backend. The query has [queryType](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudwatch/cloudwatch.go#L285-L286) for distinguishing these queries from for example time series queries, and then each metric find query has a [subType](https://github.com/grafana/grafana/blob/main/pkg/tsdb/cloudwatch/metric_find_query.go#L265-L283) to distinguish the meta data query types. This is not the idomatic way to use the plugin api. For meta data requests of this kind, the backend should expose a CallResourceHandler. The backend can then expose a mux with routes for all the metric find query sub types. </div>

### PR DESCRIPTION
**Pull request by iwysiu**
**What this PR does / why we need it**:
Uses the callResouceHandler api instead of the QueryData api.

**Which issue(s) this PR fixes**:

Fixes #40748 

**Special notes for your reviewer**:

### CODE REVIEW

**Review by wbrowne — APPROVED** [review]
LGTM

File: pkg/tsdb/cloudwatch/resource_handler.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+func handleResourceReq(handleMetricFind handleFn) func(rw http.ResponseWriter, req *http.Request) {
+	return func(rw http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		pluginContext := httpadapter.PluginConfigFromContext(ctx)
+		err := req.ParseForm()
+		if err != nil {
+			writeResponse(rw, http.StatusBadRequest, fmt.Sprintf("unexpected error %v", err))
+		}
+		data, err := handleMetricFind(pluginContext, req.URL.Query())
```
Comment (andresmgot) [inline pkg/tsdb/cloudwatch/resource_handler.go:37]: I'd avoid the "metricFind" naming so it's not confusing

File: public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts [hunk 2]
```diff
@@ -511,11 +511,11 @@ describe('CloudWatchDatasource', () => {
       describe('and region param is left out', () => {
         it('should use the default region', async () => {
           const { ds, instanceSettings } = getTestContext();
-          ds.doMetricQueryRequest = jest.fn().mockResolvedValue([]);
+          ds.doMetricResourceRequest = jest.fn().mockResolvedValue([]);
 
           await ds.metricFindQuery('metrics(testNamespace)');
```
Comment (andresmgot) [inline public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts:516]: I believe you can now remove the metricFindQuery method from `datasource.ts`? I don't see it being used anywhere.

**Review by andresmgot — APPROVED** [review]
LGTM! Just minor comments.

File: pkg/plugins/plugincontext/plugincontext.go [hunk 3]
```diff
@@ -44,10 +44,10 @@ type Provider struct {
 	logger                log.Logger
 }
 
-// Get allows getting plugin context by its ID. If datasourceUID is not empty string
+// Get allows getting plugin context by its ID. If datasource is not nil
 // then PluginContext.DataSourceInstanceSettings will be resolved and appended to
 // returned context.
-func (p *Provider) Get(ctx context.Context, pluginID string, datasourceUID string, user *models.SignedInUser, skipCache bool) (backend.PluginContext, bool, error) {
+func (p *Provider) Get(ctx context.Context, pluginID string, datasource *models.DataSource, user *models.SignedInUser, skipCache bool) (backend.PluginContext, bool, error) {
```
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/plugins/plugincontext/plugincontext.go]: Hmm. You're sure that the test doesn't provide all the needed setup/data?

**Review by marefr — CHANGES_REQUESTED** [review]
I think the change of plugin context provider is not needed. Prove me wrong

File: pkg/plugins/plugincontext/plugincontext.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/plugins/plugincontext/plugincontext.go]: The reason that this worked for the metric handler is that the metric handler [passes the datasource object](https://github.com/grafana/grafana/blob/163b570f5ded71760c1551b87013d1e29ac20839/pkg/services/query/query.go#L225) to build the [instance settings](https://github.com/grafana/grafana/blob/163b570f5ded71760c1551b87013d1e29ac20839/pkg/services/query/query.go#L122) while the resource handler used to just pass the [datasourceID](https://github.com/grafana/grafana/blob/163b570f5ded71760c1551b87013d1e29ac20839/pkg/plugins/plugincontext/plugincontext.go#L50) which in this case, was nil. Changing the function brings the resource handler in line with the metric handler, which made more sense to me.

File: pkg/tsdb/cloudwatch/resource_handler.go (hunk 1 above)
Comment (iwysiu) [reply pkg/tsdb/cloudwatch/resource_handler.go]: done

File: public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts (hunk 2 above)
Comment (iwysiu) [reply public/app/plugins/datasource/cloudwatch/specs/datasource.test.ts]: done

File: pkg/plugins/plugincontext/plugincontext.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/plugins/plugincontext/plugincontext.go]: > which in this case, was ""

That specific thing feels like something not setup as expected in the test? 

> Changing the function brings the resource handler in line with the metric handler, which made more sense to me. If a valid models.Datasource can be created without a datasourceUID (which was true before this pr), they should still get added correctly to plugin contexts.

I see. 

In addition, feels like [CallDatasourceResource](https://github.com/grafana/grafana/blob/b4dd3138ced9268e4e4fb8d09dc7c80ada2d7ee6/pkg/api/datasources.go#L480) maybe shouldn't use [callPluginResource](https://github.com/grafana/grafana/blob/d9a5af3d6f1d6fb37866a46d36c1b57aaa981149/pkg/api/plugins.go#L478). Compare with [checkDatasourceHealth](https://github.com/grafana/grafana/blob/8bb3de3037f9b78e54285a3db7d190d4c22a84b0/pkg/api/datasources.go#L517). @wbrowne any thoughts?

[the remark this answers is not in the document]
Comment (iwysiu) [reply pkg/plugins/plugincontext/plugincontext.go]: Ok. Done

**Review by wbrowne — APPROVED** [review]

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**